### PR TITLE
Fixed bug predicting with xgboost + handling NAs in training data

### PR DESCRIPTION
**Pull request by martinju**
This is my first pull request ever so please bear with me (and let me know) if I am doing something wrong.

This tiny pull request does 2 things: 

1. It fixes what I believe is a bug in predict_model.xgb.Booster which does not transfer the input to a matrix before calling predict (which xgboost requires to do prediction). I simply did this transformation, which makes xgboost work with lime.
2. It allows NA values to be included in the training data when calling lime(). Methods like xgboost can both train and predict even if there are NA values involved, which should then also be OK for lime. The problem in the master was that quantiles and ranges could not be computed when training data passed to lime() included NA-values. This pull request just removes the NA values from this bin/range computations. NOTE: Lime still cannot explain predictions including NA values. I tried to look into this as well, but did find a clean solution. One should probably modifying the input to glmnet() and glm.fit().

Test scripts which shows the working fork and error from the master is given below. Note that the main.package.dir and devel.package.dir needs to be modified appropriately for others to use it. (Please let me know if there are better ways to compare and test my fork vs the master.)

```
# Example running xgboost, handling NA in training data when calling lime()

library(xgboost)
library(devtools)

# Install main and my fork of the lime package
main.package.dir <- "C:\\Users\\jullum\\Documents\\R\\win-library\\3.4"
devel.package.dir <- "C:\\Users\\jullum\\Documents\\R\\win-library\\devel_packages"
withr::with_libpaths(new = main.package.dir, install_github('thomasp85/lime'))
withr::with_libpaths(new = devel.package.dir, install_github('martinju/lime'))

# Defining data 
set.seed(123)
x = matrix(rnorm(100*10),ncol=10)
y = round(runif(100))
data.to.explain <- as.data.frame(head(x))


#### Example running plain xgboost

# Running xgboost
mod = xgboost(data = x,label=y,nrounds = 10,objective = "binary:logistic")

# My fork of the lime package. Works
library(lime, lib.loc = devel.package.dir)
explainer <- lime(x=as.data.frame(x),model=mod)
explanation <- explain(data.to.explain, explainer, labels = 1, n_features = 5)
plot_features(explanation)

# The master version fails
detach("package:lime", unload=TRUE)
library(lime,lib.loc = main.package.dir)
explanation <- explain(data.to.explain, explainer, labels = 1, n_features = 5)
#Error in xgb.DMatrix(newdata, missing = missing) : 
#  xgb.DMatrix does not support construction from list

#### Example runing xgboost with NA values in training

# Modify data
x.NA <- x
x.NA[,1] = ifelse(x.NA[,2] > 0, NA, x.NA[,1]) # introduce random NAs in V1


# Running xgboost
mod.NA = xgboost(data = x.NA,label=y,nrounds = 10,objective = "binary:logistic")

# My fork of the lime package. Works
detach("package:lime", unload=TRUE)
library(lime, lib.loc = devel.package.dir)
explainer <- lime(x=as.data.frame(x.NA),model=mod.NA)
explanation <- explain(data.to.explain, explainer, labels = 1, n_features = 5)
plot_features(explanation)

# The master version fails
detach("package:lime", unload=TRUE)
library(lime,lib.loc = main.package.dir)
explainer <- lime(x=as.data.frame(x.NA),model=mod.NA)
#Error in quantile.default(x[[i]], seq(0, 1, length.out = n_bins + 1)) : 
#  missing values and NaN's not allowed if 'na.rm' is FALSE
```

### CODE REVIEW

File: R/dataframe.R [hunk 1]
```diff
@@ -38,10 +38,10 @@ lime.data.frame <- function(x, model, bin_continuous = TRUE, n_bins = 4, quantil
   explainer$bin_cuts <- setNames(lapply(seq_along(x), function(i) {
     if (explainer$feature_type[i] %in% c('numeric', 'integer')) {
       if (quantile_bins) {
-        bins <- quantile(x[[i]], seq(0, 1, length.out = n_bins + 1))
+        bins <- quantile(x[[i]], seq(0, 1, length.out = n_bins + 1),na.rm=T)
         bins[!duplicated(bins)]
       } else {
-        d_range <- range(x[[i]])
+        d_range <- range(x[[i]],na.rm=T)
```
Comment (thomasp85) [inline R/dataframe.R]: I have the same fix locally - just remove your modifications here and lets focus this PR on xgboost support :-)

**Comment by thomasp85** [comment]
Thanks for this PR 🙂

Would it not make sense to convert the data.frame into the native `xgboost` storage class, rather than a matrix? A matrix will coerce all columns into the same native type...

Also, can you look into why the vignette does not compile? I suspect your changes has some interference with the standard text-analysis explanations (probably that it forces sparse matrices into a dense format)

**Comment by martinju** [comment]
OK, so I reverted the changes in dataframe.R (you do realize that there are 2 places where na.rm=T are needed, right?). 

I also figured out the problem with the vignette compiling. predict_xgb.Booster is also used by the text version and in that case newdata is a dgCMatrix which then shouldn't be coerced to a non-sparse matrix. I therefore added a check whether newdata is a matrix and only in that case convert to a matrix.

Regarding what to convert the data.frame to I think a matrix is an OK choice since both other options which xgboost handles (dgCMatrix and xgb.DMatrix) requires the input of them to be a matrix anyway (it cannot take a data.frame directly). Internally xgboost converts the matrix to a xgb.DMatrix anyway so there should be no speed differences. The alternatives are something like 
`newdata <- xgb.DMatrix(data=as.matrix(newdata))` 
and 
`newdata <- Matrix(as.matrix(newdata),sparse=T)`. 
If you think either of those are more natural I can change that.

**Comment by pommedeterresautee** [comment]
Is it possible that there is a memory issue with a very large sparse matrix?

**Comment by martinju** [comment]
Possibly, but I think you would seldom pass so many explanations that this part of the procedure really would matter memory-wise. However, predict.xgb.Booster does this with its input: 
 if (!inherits(newdata, "xgb.DMatrix"))  newdata <- xgb.DMatrix(newdata, missing = missing)

So, doing newdata <- xgb.DMatrix(as.matrix(newdata)) might the way to go to avoid a matrix copy of the newdata. Just let me know what you think, and I will change it.

**Comment by thomasp85** [comment]
Remember `predict_model()` receives all permutations in addition to the observations so 10 requested explanations expands to 50000 predictions with default settings. Thus it should definitely be taken hand of

```r
newdata <- xgb.DMatrix(as.matrix(newdata))
```

Will also coerce to an intermediary matrix so this comes with the same limitations. Is it not possible to use the sparse structures from the `Matrix` package?

**Comment by martinju** [comment]
Yes, good point :) 
But all other methods store the permutations in a data.frame, right?

You can use sparse structures, but if the permutations needs to first be stored as a data.frame, then a matrix, then a sparse Matrix, and then finally a xgb.DMatrix, I am not sure about the gain...

As newdata is a data.frame you would as far as I know need to do
`newdata <- xgb.DMatrix(Matrix(as.matrix(newdata),sparse=T))
`
I think the only way around intermediate inefficient object types is to generate the permutations directly in a sparse structure.

**Comment by thomasp85** [comment]
This is only true if you're thinking about explaining tabular data - if you're working with text data then `predict_model` will receive whatever comes out of the preprocessing function (often a sparse document-term matrix)

I think the best approach is probably to check if `newdata` is `data.frame` and only in that case convert to `matrix`, prior to coercing to `xib.DMatrix`

**Comment by martinju** [comment]
Ok. Checking for data.frame format before conversion was already in place. I added the xgb.DMatrix call. I hope this is ready for merging now :-)

**Comment by thomasp85** [comment]
This looks good - thanks

**Pull request merged**